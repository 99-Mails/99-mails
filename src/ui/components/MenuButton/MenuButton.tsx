import { ChevronDownIcon } from "@chakra-ui/icons";
import {
  Menu,
  MenuList,
  MenuButton,
  MenuButtonProps,
  forwardRef,
  ButtonGroup,
  Button,
  ButtonProps,
  MenuOptionGroup,
  MenuItemOption,
} from "@chakra-ui/react";
import { useState } from "react";

export interface MenuWrapperProps extends MenuButtonProps {
  cb: Fn;
  onChange: Fn;
  data: any[];
}

const MenuWrapper = forwardRef<MenuWrapperProps, "button">((props, ref) => {
  const { cb, onChange, data, ...rest } = props;

  return (
    <Menu
      isLazy
      eventListeners={true}
      flip={true}
      preventOverflow={true}
      boundary="clippingParents"
      placement="auto-end"
      strategy="absolute"
    >
      <MenuButton ref={ref} {...rest}>
        <ChevronDownIcon />
      </MenuButton>
      <MenuList>
        <MenuOptionGroup onChange={(e) => onChange(e)}>
          {data.map((item, index) => (
            <MenuItemOption
              data-testid="menu-wrapper-button"
              key={index}
              value={item}
              onClick={() => cb(item)}
            >
              {item}
            </MenuItemOption>
          ))}
        </MenuOptionGroup>
      </MenuList>
    </Menu>
  );
});

export interface MenuButtonContainerProps extends ButtonProps {
  initialValue: string;
  onChange: Fn;
  data: any[];
}

const MenuButtonContainer = forwardRef<MenuButtonContainerProps, "button">(
  (props, ref) => {
    const { initialValue, data, onChange, ...restProps } = props;
    const [value, setValue] = useState(initialValue);

    return (
      <ButtonGroup isAttached variant="outline">
        <Button data-testid="menu-button" ref={ref} {...restProps} size="sm">
          + {value}
        </Button>
        <MenuWrapper
          disabled={false}
          data={data}
          cb={setValue}
          onChange={onChange}
          data-testid="menu-wrapper"
        />
      </ButtonGroup>
    );
  }
);

export { MenuButtonContainer };
export { MenuWrapper };
