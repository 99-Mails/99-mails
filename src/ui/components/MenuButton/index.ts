export { MenuButtonContainer as default } from "./MenuButton";
export type { MenuButtonContainerProps } from "./MenuButton";
export * from "./theme";
