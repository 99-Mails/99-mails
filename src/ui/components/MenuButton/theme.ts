import { menuAnatomy } from "@chakra-ui/anatomy";
import { createMultiStyleConfigHelpers } from "@chakra-ui/react";

const { definePartsStyle, defineMultiStyleConfig } =
  createMultiStyleConfigHelpers(menuAnatomy.keys);

const baseStyle = definePartsStyle({
  button: {
    px: 2,
    transition: "all 0.2s",
    borderRadius: "md",
    borderWidth: "1px",
    _hover: {
      bg: "gray.400",
    },
    _focus: { boxShadow: "outline" },
  },
});

export const menuTheme = defineMultiStyleConfig({ baseStyle });
