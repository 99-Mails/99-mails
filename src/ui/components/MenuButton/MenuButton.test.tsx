import { vi, it, expect } from "vitest";
import { render, screen, userEvent } from "../../../tests/test-utils";
import { MenuButtonContainer } from "./MenuButton";

it("should render correctly", () => {
  const fruitList = ["apple", "banana", "orange"];
  const selectedFruit = vi.fn();

  const { asFragment } = render(
    <MenuButtonContainer
      data={fruitList}
      initialValue="apple"
      onChange={selectedFruit}
    />
  );
  expect(asFragment()).toMatchSnapshot();
});

it("should show the initial value correctly", async () => {
  const fruitList = ["apple", "banana", "orange"];
  const selectedFruit = vi.fn();

  render(
    <MenuButtonContainer
      data={fruitList}
      initialValue="apple"
      onChange={selectedFruit}
    />
  );

  expect(await screen.findByTestId("menu-button")).toHaveTextContent("apple");
});

// TODO: fix vitest console error
// When testing, code that causes React state updates should be wrapped into act(...):

// act(() => {
//   /* fire events that update state */
// });
// /* assert on the output */

// This ensures that you're testing the behavior the user would see in the browser. Learn more at https://reactjs.org/link/wrap-tests-with-act
//     at Menu (file:///Users/mahdi/Projects/temp-email/node_modules/@chakra-ui/menu/dist/chunk-6BD5HRZF.mjs:22:11)
//     at /Users/mahdi/Projects/temp-email/src/ui/components/MenuButton/MenuButton.tsx:10:11
//     at div
//     at Styled(div) (/Users/mahdi/Projects/temp-email/node_modules/@emotion/react/dist/emotion-element-b63ca7c6.cjs.dev.js:43:23)
//     at ChakraComponent (file:///Users/mahdi/Projects/temp-email/node_modules/@chakra-ui/system/dist/chunk-E32MZNBA.mjs:42:35)
//     at ButtonGroup2 (file:///Users/mahdi/Projects/temp-email/node_modules/@chakra-ui/button/dist/chunk-PEYICJIL.mjs:36:7)
//     at /Users/mahdi/Projects/temp-email/src/ui/components/MenuButton/MenuButton.tsx:42:13
// Warning: An update to MenuItemOption inside a test was not wrapped in act(...).
it("should show the menu when it's clicked", async () => {
  const user = userEvent.setup();

  const fruitList = ["apple", "banana", "orange"];
  const selectedFruit = vi.fn();

  const { container } = render(
    <MenuButtonContainer
      data={fruitList}
      initialValue="apple"
      onChange={selectedFruit}
    />
  );

  const MenuWrapper = await screen.findByTestId("menu-wrapper");

  await user.click(MenuWrapper);

  fruitList.forEach((fruit) => {
    expect(container).toHaveTextContent(fruit);
  });
});

it("should change the value of button when an item in menu is clicked", async () => {
  const user = userEvent.setup();

  const fruitList = ["apple"];
  const selectedFruit = vi.fn();

  const { container } = render(
    <MenuButtonContainer
      data={fruitList}
      initialValue="banana"
      onChange={selectedFruit}
    />
  );

  const MenuWrapper = await screen.findByTestId("menu-wrapper");

  await user.click(MenuWrapper);

  const MenuWrapperItem = await screen.findByTestId("menu-wrapper-button");

  await user.click(MenuWrapperItem);

  expect(await screen.findByTestId("menu-button")).toHaveTextContent("apple");
});
