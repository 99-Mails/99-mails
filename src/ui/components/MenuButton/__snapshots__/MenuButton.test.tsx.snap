// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-button__group css-1ayoy1w"
    data-attached=""
    data-orientation="horizontal"
    role="group"
  >
    <button
      class="chakra-button css-1fysbuj"
      data-testid="menu-button"
      type="button"
    >
      + apple
    </button>
    <button
      aria-controls="menu-list-:r1:"
      aria-expanded="false"
      aria-haspopup="menu"
      class="chakra-menu__menu-button css-1eqi7qs"
      data-testid="menu-wrapper"
      id="menu-button-:r1:"
    >
      <span
        class="css-xl71ch"
      >
        <svg
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </span>
    </button>
    <div
      class="css-0"
      style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
    >
      <div
        aria-orientation="vertical"
        class="chakra-menu__menu-list css-sjt5nk"
        id="menu-list-:r1:"
        role="menu"
        style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
        tabindex="-1"
      />
    </div>
  </div>
</DocumentFragment>
`;
